### PR TITLE
[TEVA-1237] Homepage SEO improvements - Location and popular categories facets

### PR DESCRIPTION
**Pull request by josephhull676**
### Jira ticket URL

https://dfedigital.atlassian.net/browse/TEVA-1237

### Changes in this PR:

- Created service to cache and retrieve vacancy category facets in/from Redis
- Created initializer for vacancy facets
- Changed location category links to include counts and added popular categories (job role and subject) and counts to homepage

Screenshot 2020-10-19 at 12 19 40

### CODE REVIEW

**Comment by alexbowen** [comment]
need a description to this

File: app/controllers/vacancies_controller.rb [hunk 1]
```diff
@@ -40,6 +40,7 @@ def params
 
   def algolia_search_params
     strip_empty_checkboxes(:jobs_search_form, %i[job_roles phases working_patterns])
+    params['job_roles'] = [params['job_roles']] if params['job_roles'].is_a?(String)
```
Comment (cpjmcquillan) [inline app/controllers/vacancies_controller.rb:43]: Why is this necessary?

File: app/services/vacancy_facets.rb [hunk 2]
```diff
@@ -0,0 +1,66 @@
+class VacancyFacets
+  FIELDS = %i[job_roles subjects cities counties].freeze
+
+  def initialize(store: Redis.new)
+    @store = store
+  end
+
+  def get(field)
+    return {} if !FIELDS.include?(field) || !store.exists?(field)
```
Comment (cpjmcquillan) [inline app/services/vacancy_facets.rb:9]:
```ruby
return {} unless FIELDS.include?(field) && store.exists?(field)
```
Might be more readable/ruby, but I don't mind either

File: app/services/vacancy_facets.rb [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  def subject_facet(subjects)
+    subjects.filter_map { |subject|
+      count = algolia_facet_count(keyword: subject.first)
+      [subject.first, count] if count.positive?
+    }.to_h
+  end
+
+  def location_category_facet(locations)
+    locations.filter_map { |location|
+      count = algolia_facet_count(location: location, location_category: location, radius: 10)
```
Comment (cpjmcquillan) [inline app/services/vacancy_facets.rb:57]: You should be able to provide just `location: location` or `location_category: location` only

File: app/services/vacancy_facets.rb [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  def set_cities_facets
+    store.set(:cities, location_category_facet(CITIES).to_json)
+  end
+
+  def set_counties_facets
+    store.set(:counties, location_category_facet(COUNTIES).to_json)
+  end
+
+  def job_role_facet(job_roles)
+    job_roles.filter_map { |job_role|
```
Comment (cpjmcquillan) [inline app/services/vacancy_facets.rb:42]: Sometimes I find it clearer to explicitly include the unused parameter like `|job_role, _|` when it's a hash but we only care about one of the key/value

File: spec/services/vacancy_facets_spec.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
+require 'rails_helper'
```
Comment (cpjmcquillan) [inline spec/services/vacancy_facets_spec.rb:1]: A new line under the imports helps readability I find (and you did it below too)

File: app/views/pages/home/_vacancy_facets.haml [hunk 6]
```diff
@@ -0,0 +1,30 @@
+.govuk-grid-row
+  .govuk-grid-column-one-half
+    %h3.govuk-heading-s= t('.cities')
+    %ul.govuk-list.two-column-list{ aria: { label: "#{t('.cities')} list" } }
+      - @cities_facets.first(20).each do |city, count|
```
Comment (cpjmcquillan) [inline app/views/pages/home/_vacancy_facets.haml:5]: Are these sorted by number of hits?

File: spec/services/vacancy_facets_spec.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+      it 'returns an empty hash' do
+        expect(subject.get(field)).to eq({})
+      end
+    end
+  end
+
+  describe '#refresh' do
+    it 'calls Search::VacancySearchBuilder' do
+      JOB_ROLES_SUBJECTS_CITIES_AND_COUNTIES = 229
```
Comment (cpjmcquillan) [inline spec/services/vacancy_facets_spec.rb:29]: This seems a bit flaky

File: app/controllers/vacancies_controller.rb (hunk 1 above)
Comment (josephhull676) [reply app/controllers/vacancies_controller.rb:43]: When clicking on a job role facet link the value of "job_roles' in the parameters is a string. As string are not permitted for job_roles. We need to convert it to an array I believe. Perhaps there's a better way of doing this? 
```
.permit(:keyword, :location, :location_category, :radius, :jobs_sort, :page,
              job_roles: [], phases: [], working_patterns: [])
```

File: app/services/vacancy_facets.rb (hunk 2 above)
Comment (josephhull676) [reply app/services/vacancy_facets.rb:9]: Yup that makes sense

File: app/controllers/vacancies_controller.rb (hunk 1 above)
Comment (cpjmcquillan) [reply app/controllers/vacancies_controller.rb:43]: can you add `:job_roles` as an argument in the permit?

File: app/views/pages/home/_vacancy_facets.haml (hunk 6 above)
Comment (josephhull676) [reply app/views/pages/home/_vacancy_facets.haml:5]: No. It made sense to me to sort them by number of hits and then alphabetically, but Chris said he just wants them to be alphabetically sorted.

File: app/services/vacancy_facets.rb [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  def subject_facet(subjects)
+    subjects.filter_map { |subject|
+      count = algolia_facet_count(keyword: subject.first)
+      [subject.first, count] if count.positive?
+    }.to_h
+  end
+
+  def location_category_facet(locations)
+    locations.filter_map { |location|
+      count = algolia_facet_count(location: location, radius: 10)
```
Comment (cpjmcquillan) [inline app/services/vacancy_facets.rb:57]: Radius isn't necessary because these are location category (polygon) searches

File: app/services/vacancy_facets.rb [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    store.set(:cities, location_category_facet(CITIES).to_json)
+  end
+
+  def set_counties_facets
+    store.set(:counties, location_category_facet(COUNTIES).to_json)
+  end
+
+  def job_role_facet(job_roles)
+    job_roles.filter_map { |job_role|
+      count = algolia_facet_count(job_roles: job_role)
```
Comment (cpjmcquillan) [inline app/services/vacancy_facets.rb:43]: We should be passing in an array of job_roles to the search builder

**Review by cpjmcquillan — CHANGES_REQUESTED** [review]

File: app/views/pages/home/_vacancy_facets.haml (hunk 6 above)
Comment (cpjmcquillan) [reply app/views/pages/home/_vacancy_facets.haml:5]: Hmm. I'm not sure about that - surely we need the top 20 cities with the most results sorted alphabetically?

File: app/frontend/styles/base/_utilities.scss [hunk 10]
```diff
@@ -73,6 +73,14 @@ b {
   }
 }
 
+.two-column-list {
+
+  column-count: 2;
+  @include govuk-media-query($until: tablet) {
```
Comment (alexbowen) [inline app/frontend/styles/base/_utilities.scss:79]: you dont need this media query as column-count is already 2?

File: app/views/pages/home/_vacancy_facets.haml (hunk 6 above)
Comment (alexbowen) [reply app/views/pages/home/_vacancy_facets.haml:5]: instead of the first(20) part shouldnt this be done elsewhere in a presenter or similar or better still when the data is queried?

File: app/services/vacancy_facets.rb (hunk 9 above)
Comment (josephhull676) [reply app/services/vacancy_facets.rb:43]: `job_role` is an array. I'm not too sure what you mean?

File: app/frontend/styles/base/_utilities.scss (hunk 10 above)
Comment (josephhull676) [reply app/frontend/styles/base/_utilities.scss:79]: Make sense!

File: app/services/vacancy_facets.rb (hunk 8 above)
Comment (josephhull676) [reply app/services/vacancy_facets.rb:57]: Changed

File: app/services/vacancy_facets.rb (hunk 9 above)
Comment (josephhull676) [reply app/services/vacancy_facets.rb:43]: Have changed this now.

File: spec/services/vacancy_facets_spec.rb (hunk 7 above)
Comment (josephhull676) [reply spec/services/vacancy_facets_spec.rb:30]: Changed this now. See what you think after iv'e pushed

File: app/views/pages/home/_vacancy_facets.haml (hunk 6 above)
Comment (josephhull676) [reply app/views/pages/home/_vacancy_facets.haml]: Was thinking I should probably bring this up at stand up.

Comment (josephhull676) [reply app/views/pages/home/_vacancy_facets.haml]: Have now changed it so it's sorted by count and then alphabetically.

File: app/controllers/pages_controller.rb [hunk 11]
```diff
@@ -3,6 +3,11 @@ class PagesController < ApplicationController
 
   def home
     @jobs_search_form = VacancyAlgoliaSearchForm.new
+    @vacancy_facets = VacancyFacets.new
+    @job_roles_facets = @vacancy_facets.get(:job_roles)
+    @subjects_facets = @vacancy_facets.get(:subjects).first(10)
```
Comment (cesidio) [inline app/controllers/pages_controller.rb:8]: `.first(10)` happens already in the service. same for the following lines

File: app/services/vacancy_facets.rb [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+  def sort_and_limit(facet, number_of_results)
+    facet.sort_by { |_, count| -count }.first(number_of_results).sort.to_h
+  end
+
+  def job_role_facet
+    Vacancy::JOB_ROLE_OPTIONS.keys.each_with_object({}) { |job_role, facets| facets[job_role] = algolia_facet_count(job_roles: [job_role]) }
+  end
+
+  def subject_facet
+      SUBJECT_OPTIONS.each_with_object({}) { |subject, facets| facets[subject.first] = algolia_facet_count(keyword: subject.first) }
```
Comment (cesidio) [inline app/services/vacancy_facets.rb:34]: 4 space indentation

File: app/controllers/pages_controller.rb (hunk 11 above)
Comment (josephhull676) [reply app/controllers/pages_controller.rb:8]: I thought I had removed this. My mistake!

File: app/services/vacancy_facets.rb (hunk 12 above)
Comment (josephhull676) [reply app/services/vacancy_facets.rb]: Changes

**Review by cpjmcquillan — APPROVED** [review]

**Review by cpjmcquillan — COMMENTED** [review]
Looks good to me, and very nice work 👍 🥇 🍾
